### PR TITLE
Fix MSVC template demangling with ref to mangled symbol

### PR DESCRIPTION
**Pull request by GustavoLCR**
**Your checklist for this pull request**
- [x] I've read the [guidelines for contributing](https://github.com/radareorg/radare2/blob/master/DEVELOPERS.md) to this repository
- [x] I made sure to follow the project's [coding style](https://github.com/radareorg/radare2/blob/master/DEVELOPERS.md#code-style)
- [x] I've added tests that prove my fix is effective or that my feature works (if possible)
- [ ] I've updated the documentation and the [radare2 book](https://github.com/radareorg/radare2book) with the relevant information (if needed)

**Detailed description**

* Also fix not adding the template name as an abbreviation
* Add test

**Test plan**

Tests pass

### CODE REVIEW

**Review by ret2libc — APPROVED** [review]
Honestly, I haven't looked at each line, but... the changes look reasonable, tests pass and there is a new test for the fixed case, so I'm pretty much ok with this :)

File: libr/bin/mangling/microsoft_demangle.c [hunk 1]
```diff
@@ -109,8 +109,10 @@ static void init_state_struct(SStateInfo *state, char *buff_for_parsing);
 static EDemanglerErr get_type_code_string(char *sym, unsigned int *amount_of_read_chars, char **str_type_code);
 static int init_type_code_str_struct(STypeCodeStr *type_coder_str);
 static void free_type_code_str_struct(STypeCodeStr *type_code_str);
-static int get_template(char *buf, SStrInfo *str_info);
+static int get_template(char *buf, SStrInfo *str_info, bool memorize);
 static char *get_num(SStateInfo *state);
+static EDemanglerErr parse_data_type(char *sym, char **demangled_type, char **access_modifier, int *len, bool *is_static, bool *is_implicit_this_pointer);
+int get_namespace_and_name(char *buf, STypeCodeStr *type_code_str, unsigned int *amount_of_names, bool memorize);
```
Comment (trufae) [inline libr/bin/mangling/microsoft_demangle.c:115]: this should be static i think

Comment (trufae) [reply libr/bin/mangling/microsoft_demangle.c:115]: use size_t instead of unsigned int

File: libr/bin/mangling/microsoft_demangle.c [hunk 2]
```diff
@@ -488,8 +524,8 @@ static int get_template(char *buf, SStrInfo *str_info) {
 /// \param amount_of_names Amount of names that was in list
 /// \return Return amount of processed chars
 ///
-int get_namespace_and_name(	char *buf, STypeCodeStr *type_code_str,
-							int *amount_of_names)
+int get_namespace_and_name(char *buf, STypeCodeStr *type_code_str,
```
Comment (trufae) [inline libr/bin/mangling/microsoft_demangle.c:527]: not sure if char *buf can be const char *buf, and also the return could be bool

File: libr/bin/mangling/microsoft_demangle.c [hunk 3]
```diff
@@ -1207,137 +1243,92 @@ static EDemanglerErr get_type_code_string(char *sym, unsigned int *amount_of_rea
...
 			ptr64 = "__ptr64";
 			curr_pos++;
 		}
 
-	#define SET_STORAGE_CLASS(letter, storage_class_str) { \
-		case letter: \
-			storage_class = storage_class_str; \
-			break; \
+#define SET_STORAGE_CLASS(letter, storage_class_str) \
+	case letter: {                                   \
```
Comment (trufae) [inline libr/bin/mangling/microsoft_demangle.c:1291]: dont indent the backslashes

File: libr/bin/mangling/microsoft_demangle.c [hunk 4]
```diff
@@ -1207,137 +1243,92 @@ static EDemanglerErr get_type_code_string(char *sym, unsigned int *amount_of_rea
...
-		access_modifier = modifier_str; \
-		(flag_set) = 1; \
-		break; \
+	case '6': // compiler generated static
+	case '7': // compiler generated static
+		return eDemanglerErrUnsupportedMangling;
+
+#define SET_ACCESS_MODIFIER(letter, flag_set, modifier_str)   \
+	case letter: {                                            \
+		if (access_modifier) *access_modifier = modifier_str; \
```
Comment (trufae) [inline libr/bin/mangling/microsoft_demangle.c:1329]: add braces around the if statements, its not needed for the case

File: libr/bin/mangling/microsoft_demangle.c [hunk 5]
```diff
@@ -1358,18 +1349,83 @@ static EDemanglerErr parse_microsoft_mangled_name(char *sym, char **demangled_na
...
+/// <public name> ::= ?<name>@[<namespace>@](0->inf)@3<type><storage class>
+/// mangled name of a static class member object:
+/// <public name> ::= ?<name>@[<classname>@](1->inf)@2<type><storage class>
+///////////////////////////////////////////////////////////////////////////////
+static EDemanglerErr parse_microsoft_mangled_name(char *sym, char **demangled_name) {
+	STypeCodeStr type_code_str;
+	STypeCodeStr func_str;
+	EDemanglerErr err = eDemanglerErrOK;
+
+	unsigned int is_abbr_type = 0;
```
Comment (trufae) [inline libr/bin/mangling/microsoft_demangle.c:1376]: size_t

File: libr/bin/mangling/microsoft_demangle.c [hunk 6]
```diff
@@ -1358,18 +1349,83 @@ static EDemanglerErr parse_microsoft_mangled_name(char *sym, char **demangled_na
...
+	return eDemanglerErrOK;
+}
+
+///////////////////////////////////////////////////////////////////////////////
+/// public mangled name of global object:
+/// <public name> ::= ?<name>@[<namespace>@](0->inf)@3<type><storage class>
+/// mangled name of a static class member object:
+/// <public name> ::= ?<name>@[<classname>@](1->inf)@2<type><storage class>
+///////////////////////////////////////////////////////////////////////////////
+static EDemanglerErr parse_microsoft_mangled_name(char *sym, char **demangled_name) {
```
Comment (trufae) [inline libr/bin/mangling/microsoft_demangle.c:1371]: const char *sym i think

File: libr/bin/mangling/microsoft_demangle.c [hunk 7]
```diff
@@ -1358,18 +1349,83 @@ static EDemanglerErr parse_microsoft_mangled_name(char *sym, char **demangled_na
...
+		err = eDemanglerErrMemoryAllocation;
+		goto parse_microsoft_mangled_name_err;
+	}
+
+	if (!init_type_code_str_struct (&type_code_str)) {
+		err = eDemanglerErrMemoryAllocation;
+		goto parse_microsoft_mangled_name_err;
+	}
+
+	len = get_namespace_and_name (curr_pos, &type_code_str, &i, false);
```
Comment (trufae) [inline libr/bin/mangling/microsoft_demangle.c:1409]: assign and define in the same line

**Review by trufae — CHANGES_REQUESTED** [review]

**Review by XVilka — APPROVED** [review]

**Pull request merged**